core/src/main/java/com/linecorp/armeria/common/QueryStringDecoder.java

```diff
@@ -0,0 +1,195 @@
+/*
+ * Copyright 2019 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+/*
+ * Copyright 2012 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.common;
+
+import static io.netty.util.internal.StringUtil.EMPTY_STRING;
+import static io.netty.util.internal.StringUtil.SPACE;
+import static io.netty.util.internal.StringUtil.decodeHexByte;
+
+import java.nio.ByteBuffer;
+import java.nio.CharBuffer;
+import java.nio.charset.CharacterCodingException;
+import java.nio.charset.CharsetDecoder;
+import java.nio.charset.CoderResult;
+import java.nio.charset.StandardCharsets;
+
+import javax.annotation.Nullable;
+
+import com.google.common.annotations.VisibleForTesting;
+
+import io.netty.util.CharsetUtil;
+
+final class QueryStringDecoder {
+
+    // Forked from netty-4.1.43.
+    // https://github.com/netty/netty/blob/7d6d953153697bd66c3b01ca8ec73c4494a81788/codec-http/src/main/java/io/netty/handler/codec/http/QueryStringDecoder.java
+
+    @SuppressWarnings("checkstyle:FallThrough")
+    static QueryParams decodeParams(@Nullable String s, int paramsLimit, boolean semicolonIsSeparator) {
+        if (s == null) {
+            return QueryParams.of();
+        }
+
+        final int len = s.length();
+
+        // Skip the leading question marks.
```

Comment:
Just wanted to skip the initial question mark automatically, but it seems like we need to leave it up to the caller for less ambiguity.

Reply:
This is definitely a hard point. I forget, does RequestHeaders.query() return with a question mark or not? If no then we're good as is, otherwise I'd probably strip a single question mark and make clear in javadoc that input needs to have the leading question mark of a query string (or alternatively we need to make sure javadoc says the opposite).

Reply:
There's no `RequestHeaders.query()`, but `RequestContext.query()` returns without a leading `?`. Anyways, I updated not to strip a question mark. After all, it seemed too ambiguous. :sweat_smile: